core/src/Kokkos_HostSpace.hpp

```diff
@@ -205,7 +205,13 @@ struct DeepCopy<HostSpace, HostSpace, ExecutionSpace> {
     exec.fence(
         "Kokkos::Impl::DeepCopy<HostSpace, HostSpace, "
         "ExecutionSpace>::DeepCopy: fence before copy");
-    hostspace_parallel_deepcopy_async(dst, src, n);
+    // If ExecutionSpace is a Device ExecutionSpace, we should call
+    // this from DefaultHostExecutionSpace
+    if constexpr (std::is_same_v<ExecutionSpace, DefaultExecutionSpace>) {
+      hostspace_parallel_deepcopy_async(dst, src, n);
+    } else {
+      hostspace_parallel_deepcopy_async(exec, dst, src, n);
+    }
```

Comment:
We don't need to fence if we are calling the `hostspace_parallel_deepcopy_async` overload taking an execution space instance.

Reply:
Something like this? Or can I completely remove this fence since this function takes an execution space argument.

```C++
    if constexpr (std::is_same_v<ExecutionSpace, DefaultExecutionSpace>) {
      exec.fence(
          "Kokkos::Impl::DeepCopy<HostSpace, HostSpace, "
          "ExecutionSpace>::DeepCopy: fence before copy");
      hostspace_parallel_deepcopy_async(dst, src, n);
    } else {
      hostspace_parallel_deepcopy_async(exec, dst, src, n);
    }
```

Reply:
Yes, this looks good. We still need the fence since we can't enqueue the work on the given execution space instance in the first case and the deepcopy might access memory used by that execution space instance.

Reply:
More like
```C++
if (std::is_same_v<ExecutionSpace, DefaultExecutionSpace> 
  && !std::is_same_v<DefaultExecutionSpace, DefaultHostExecutionSpace>)

Reply:
If they are the same we would use the specialization in line 187 (which we could get rid of).